core/src/main/java/io/grpc/ClientCall.java

```diff
@@ -108,7 +108,8 @@
   public abstract static class Listener<T> {
 
     /**
-     * The response headers have been received. Headers always precede messages.
+     * The response headers have been received. Headers always precede messages. Since {@link
+     * Metadata} is not thread-safe, the caller must not access {@code headers} after this point.
```

Comment:
This is awkward, since move user will be receiving this callback (rather than forwarding it).   It currently reads as if the headers cannot be accessed upon getting this callback.

Reply:
I missed that, maybe split that sentence out?  Also, regardless of threading, its still not safe for callers to mutate the MD afterwards.

Reply:
You mean make a new paragraph? That'd be easy.

Part of the point is they can't even _read_ from the object. There's been some logging interceptors internally that were getting that wrong, which prompted fixing up these missed spots.

Reply:
new paragraph is fine.    I would say read or write, rather than access to be more explicit.  Those are the keywords I would look for.  Optional, but my opinion anyways.